agent/hcp/telemetry/otlp_transform.go

```diff
@@ -0,0 +1,167 @@
+package telemetry
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/go-multierror"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+	cpb "go.opentelemetry.io/proto/otlp/common/v1"
+	mpb "go.opentelemetry.io/proto/otlp/metrics/v1"
+	rpb "go.opentelemetry.io/proto/otlp/resource/v1"
+)
+
+// TransformOTLP returns an OTLP ResourceMetrics generated from OTEL metrics. If rm
+// contains invalid ScopeMetrics, an error will be returned along with an OTLP
+// ResourceMetrics that contains partial OTLP ScopeMetrics.
+func transformOTLP(rm *metricdata.ResourceMetrics) (*mpb.ResourceMetrics, error) {
+	sms, err := scopeMetrics(rm.ScopeMetrics)
+	return &mpb.ResourceMetrics{
+		Resource: &rpb.Resource{
+			Attributes: attributes(rm.Resource.Iter()),
+		},
+		ScopeMetrics: sms,
+	}, err
+}
+
+// scopeMetrics returns a slice of OTLP ScopeMetrics.
+func scopeMetrics(scopeMetrics []metricdata.ScopeMetrics) ([]*mpb.ScopeMetrics, error) {
+	var merr error
+	out := make([]*mpb.ScopeMetrics, 0, len(scopeMetrics))
+	for _, sm := range scopeMetrics {
+		ms, err := metrics(sm.Metrics)
+		if err != nil {
+			merr = multierror.Append(merr, err)
```

Comment:
Yeah so we don't want to prevent appending (as we want the good metrics in) but we want to filter bad metrics (which is done in the inner function). I've cleaned up this code to remove the error handling here since it's pretty much useless. After discussing with Chappie and Connor, we'd rather get metrics on when transformation failed, and we'd do this within the inner functions when the error occurs. I'm thinking of doing this in a follow-up PR. The alternative is to log the error, but it would be easier to obtain metrics.